glean-core/rlb/src/net/mod.rs

```diff
@@ -97,21 +98,35 @@ impl UploadManager {
                             let upload_url = format!("{}{}", inner.server_endpoint, request.path);
                             let headers: Vec<(String, String)> =
                                 request.headers.into_iter().collect();
-                            let result = inner.uploader.upload(upload_url, request.body, headers);
+                            let result =
+                                match inner.uploader.upload(upload_url, request.body, headers) {
```

Comment:
If you're concerned about maintaining the abstraction, you could drill through a return value from `PingUploadManager::process_ping_upload_response`. Then you could add this `UploadResult::Done` case to its `match`, returning `PingUploadTask::Done` or something. Then handle the correct `Done` here.

Reply:
I'm not sure I understand what you mean.
Right now `glean_process_ping_upload_response` doesn't return anything. Are you suggesting it _should_ return something that this uploading loop handles to determine the next step?
(That again doesn't seem _too_ bad I think)

Reply:
That's precisely what I meant. Have the thing that handles `UploadResult::*` handle this new one, then have it return the result all the way back out here.